### PR TITLE
Vara's Week 12 and Week 13 blog

### PR DESCRIPTION
**Pull request by lb-97**

### CODE REVIEW

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+============================================================
+
+.. post:: August 21 2023
+   :author: Vara Lakshmi Bayanagari
+   :tags: google
+   :category: gsoc
+
+What I did this week
+~~~~~~~~~~~~~~~~~~~~
+Monai's VQVAE results on T1-weighted NFBS dataset, 125 samples, for batch_size=5 were qualitatively and quantitatively superior to all previous results. I continued the same experiments on the T1-weighted CC359(Calgary-Campinas-359) public dataset consisting of 359 anatomical MRI volumes of healthy individuals. Preprocessed the data using existing `transform_img` function -
```
Comment (pjsjongsung) [inline doc/posts/2023/2023_08_21_vara_week_12_13.rst:11]: batch_size=5 -> batch size of

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+   :alt: Combined trainings plots for all experiments
+   :width: 800
+
+Inference results on the best performing model - B12-both - is as follows-
+
+.. image:: /doc/_static/vqvae-monai-B12-both.png
+   :alt: VQVAE-Monai-B12-both reconstructions & originals showing equally spaced 5 slices for 2 samples
+   :width: 800
+
+This shows that our training not only converged quickly but also improved visually. Here's a comparison of our current best performing model i.e., VQVAE-Monai-B12-both & the previous one i.e., VQVAE-Monai-B5-NFBS. The test reconstruction loss is 0.0013 & 0.0015 respectively.
```
Comment (pjsjongsung) [inline doc/posts/2023/2023_08_21_vara_week_12_13.rst:28]: I don't think it makes sense to compare losses of models that use different datasets.

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+.. image:: /doc/_static/vqvae-reconstructions-comparison.png
+   :alt: VQVAE reconstruction comparison for B12-both & B5-NFBS
+   :width: 800
+
+I also carried Diffusion Model training for the bets performing B12-both model for 300 & 500 diffusion steps and the training curve obtained is as follows-
```
Comment (pjsjongsung) [inline doc/posts/2023/2023_08_21_vara_week_12_13.rst:34]: carried -> carried out
bets -> best

File: doc/_static/dm3d-monai-training-curves.png
Comment (pjsjongsung) [inline doc/_static/dm3d-monai-training-curves.png:1]: Please provide more information on what the names of the models indicate and what the y axis is. Also, provide more information on the figures with the images. Like which dataset is the subject from, what does each row represent, etc.

**Review by sreekarchigurupati — COMMENTED** [review]
Will it be possible to rotate the output images to match the input axes? Might be better for comparison

File: doc/_static/vqvae-monai-B12-both.png
Comment (pjsjongsung) [inline doc/_static/vqvae-monai-B12-both.png:1]: Please add more comments on the results. Do they look ok? What are some details missing? Will further training improve the results? If not, what can we do? etc

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst (hunk 2 above)
Comment (lb-97) [reply doc/posts/2023/2023_08_21_vara_week_12_13.rst:28]: Both of those mentioned are test losses, it gives an idea of the performance of the model. Think of it as comparing performance between training using 2 different samples, with different lengths, from a data population.

**Comment by lb-97** [comment]
> Will it be possible to rotate the output images to match the input axes? Might be better for comparison

I don't understand, can you explain what input axes are

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst (hunk 2 above)
Comment (pjsjongsung) [reply doc/posts/2023/2023_08_21_vara_week_12_13.rst:28]: If the test losses are from a totally separate dataset, that makes sense. Is it not from CC359 or NFBS?

**Comment by sreekarchigurupati** [comment]
image
We need a view where the dorsal ventral axis is top-down

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+What is coming up next week
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Wrapping up documentation & final report
+
+Did I get stuck anywhere
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Yes, I carried out debugging to understand the generation pipeline of the Diffusion Model. Cross-checked implementations of posterior mean & variance in the code base with respective formulas from the paper, as well as with MONAI's DDPM implementation. Didn't come across any error, yet the generated samples are erroneous.
```
Comment (sreekarchigurupati) [inline doc/posts/2023/2023_08_21_vara_week_12_13.rst:50]: Please add the Diffusion result images too. It's ok if they are pure noise. Also, please document current training done for LDM (is it too less data?) (we'll start from here and increase training data in the future)

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst (hunk 2 above)
Comment (lb-97) [reply doc/posts/2023/2023_08_21_vara_week_12_13.rst:28]: VQVAE-Monai-B12-both uses both datasets(CC359,NFBS)
VQVAE-Monai-B5-NFBS uses only NFBS datasets

Additionally, the test loss for VQVAE-Monai-B10-CC, uses CC359 only, is 0.013 as well. I'll add this result as well. Hope that can help us compare results.

I'll also describe the naming convention properly in the blog.

**Comment by lb-97** [comment]
> image
> We need a midsaggital view where the dorsal ventral axis is top-down.

Sure, I can do that. But all of my past visualizations & current ones are in the same plane(anterior-posterior topdown). Rotating these pics alone would lead to confusion. Do let me know your thoughts.

File: doc/_static/vqvae-monai-B12-both.png
Comment (lb-97) [reply doc/_static/vqvae-monai-B12-both.png:1]: Sure, will do.

**Comment by sreekarchigurupati** [comment]
> > image
> > We need a midsaggital view where the dorsal ventral axis is top-down.
> 
> Sure, I can do that. But all of my past visualizations & current ones are in the same plane(anterior-posterior topdown). Rotating these pics alone would lead to confusion. Do let me know your thoughts.

It's better to use the standard axes for others to understand it better in the future. But it's fine if it's too much work to redo the visualizations

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst [hunk 5]
```diff
@@ -8,36 +8,57 @@ Finalized experiments using both datasets: Week 12 & Week13
...
+
+For future work we should look into two things - debugging Diffusion Model, scaling VQVAE model. 
+
+As a first priority, we could analyze the reason for pure noise output in DM3D generations, this would help us rule out any implementation erros(although I already did that, second pair of eyes would be helpful). Because MONAI's work shows that even without a stable convergence on their diffusion model they achieved visualizations resembling a brain. Where as our model despite achieving a stable looking training curve, our generations are no close to being convincing.
+
+As a second step, we could also try sclaing up both VQVAE as well as Diffusion Model interms of complexity, such as increasing intermediate channel dimensions from 64 to 128 or 256 as done in MONAI's 3D LDM. This hopefully may help us achieve the state-of-art on NFBS & CC359 datasets.
```
Comment (pjsjongsung) [inline doc/posts/2023/2023_08_21_vara_week_12_13.rst:61]: sclaing -> scaling
Please copy your text to word or some typo checking software to catch these.

Comment (pjsjongsung) [reply doc/posts/2023/2023_08_21_vara_week_12_13.rst:61]: Also check the spelling errors that github caught

File: doc/posts/2023/2023_08_21_vara_week_12_13.rst (hunk 2 above)
Comment (pjsjongsung) [reply doc/posts/2023/2023_08_21_vara_week_12_13.rst]: If the test loss is not from a separate dataset
1. We should call it validation loss
2. It does not make sense to compare like this. The training is biased in this case. NFBS trained models has never seen the CC359 training set, while CC359 trained ones have.

Can you clarify this on the blog post?

**Review by skoudoro — APPROVED** [review]
Thank you for the review @sreekarchigurupati and @pjsjongsung. 

I am going ahead and merge this PR. Thank @lb-97 !
